torch/_dynamo/guards.py

```diff
@@ -737,45 +776,90 @@ def compile_check_fn(
                 ("___check_tensors", check_tensors_fn),
-                ("___check_tensors_verbose", check_tensors_verbose_fn),
+                ("___check_tensors_verbose", check_tensor_verbose),
                 ("tensor_check_names", tensor_check_names),
+                ("part_map", part_map),
             ]
             + list(SYMPY_INTERP.items())
         )
         closure_vars.update(CLOSURE_VARS)
+
+        # Let's go over how this code works.
+        #
+        # 1) The cache entry structure in eval_frame.c is stored in the frame's extra field. Each cache_entry is a node
+        # in a linked list. Each cache entry represents a compiled frame with specializations. In order to know if
+        # a cache entry's compiled frame is valid for reuse, we need to invoke a function, check_fn,
+        # to compare current state against the specializations captured at compile time.
+        #
+        # 2) The function, check_fn, mentioned above, is defined by executing the function below, ___make_guard_fn
+        #
+        # 3) In this, this code is rather confusing, because it defines both ___make_guard_fn and the lambda it produces
+        # which becomes the check_fn.
+        #
+        # 4) Everything is `code` becomes the check_fn. We write it out by first defining a `__fail`, which
+        # given a code_part_id, the id() of a code_part, extract a code_part from the part_map and binds a scope
+        # to it. This is used later for evaluating the expression defined in code_part, if necessary. `__fail` is
+        # invoked if a specific sub expression of a guard is failed.
+        #
+        # 5) The sub expressions of guards are defined 1 per code_part. We iterate over the code_parts and produce
+        # code that (a) assigns the result of the expression to a variable named `passing` (b) checks if not passing,
+        # (c) and if not passing, returns __fail(code_part_id) with the id of the code_part used to produce the code.
+        # or (d) if passing, proceeds until we've run all the sub expressions through.
+        #
+        # 6) In the event that we re-enter frame evaluation having failed a guard, we return the code_part
+        # and pass it through to the frame evaluation callback. This is where downstream systems that handle guard
+        # failures hook in, like guard failure logging, or converting static shape failures to dynamic shapes (if
+        # the config is set).
+        code = "      def __fail(code_part_id): \n"
+        code += "          code_part = part_map[code_part_id] \n"
+        code += "          code_part.scope = locals() \n"
+        code += "          code_part.scope['L'] = L \n"
+        for key, value in closure_vars.items():
+            if callable(value):
+                code += f"          code_part.scope['{key}'] = {key} \n"
+        code += "          return code_part \n"
+        code += "      passing = True \n"
+        for code_part in unique(code_parts):
```

Comment:
So, you've switched to the giant sled of if branches, that would help @ysiraichi's CSE too?

Reply:
yes, it would, I think this should land before that work.